src/expenses/expensesService.js

```
@@ -0,0 +1,101 @@
+/* eslint-disable function-paren-newline */
+const { v4: uuidv4 } = require('uuid');
+
+const expenses = [];
+
+export function getAllExpenses(userId, categories, dateFrom, dateTo) {
+  let filteredExpenses = expenses;
+
+  if (typeof userId === 'number') {
+    filteredExpenses = filteredExpenses.filter((e) => e.userId === userId);
+  }
+
+  if (Array.isArray(categories)) {
+    filteredExpenses = filteredExpenses.filter((exp) =>
+      categories.includes(exp.category),
+    );
+  }
+
+  if (typeof dateFrom === 'string') {
+    filteredExpenses = filteredExpenses.filter(
+      (e) => new Date(e.spentAt) >= new Date(dateFrom),
+    );
+  }
+
+  if (typeof dateTo === 'string') {
+    filteredExpenses = filteredExpenses.filter(
+      (e) => new Date(e.spentAt) <= new Date(dateTo),
+    );
+  }
```

The date filters check `typeof dateFrom === 'string'` and `typeof dateTo === 'string'`, which will be true for query parameters, but be careful: if controllers pass `undefined` or other types the checks will fail silently. Also confirm the expected date format (ISO) and document/validate it. The date comparison logic itself (`new Date(e.spentAt) >= new Date(dateFrom)`) is OK provided `spentAt` is an ISO string and `dateFrom` is a valid date string.